File: java/src/jan/week5/boj/우주탐사선_박재환.java
package jan.week5.boj;

import java.util.*;
import java.io.*;

public class 우주탐사선_박재환 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        init(br);
        br.close();
    }

    /**
     * 모든 행성을 탐사하는데 걸리는 최소 시간을 계산한다.
     * 모든 행성을 탐사하는데 걸리는 최소 시간을 계산하여라
     */
    static final int INF = 10 * 1000 + 5;
    static int ALL;
    static int planetCount;
    static int[][] planetDist;
    static int[][] dp;
    static void init(BufferedReader br) throws IOException {
        StringTokenizer st;
        st = new StringTokenizer(br.readLine().trim());
        planetCount = Integer.parseInt(st.nextToken());
        int launchPlanet = Integer.parseInt(st.nextToken());
        ALL = (1 << planetCount) - 1;
        planetDist = new int[planetCount][planetCount];
        for(int i=0; i<planetCount; i++) {
            Arrays.fill(planetDist[i], INF);
            planetDist[i][i] = 0;
        }

        for(int i=0; i<planetCount; i++) {
            st = new StringTokenizer(br.readLine().trim());
            for(int j=0; j<planetCount; j++) {
                planetDist[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        findAllDist(planetCount);
        dp = new int[planetCount][(1 << planetCount)];
        for(int i=0; i<planetCount; i++) Arrays.fill(dp[i], -1);
        System.out.println(findMinDist(launchPlanet, (1 << launchPlanet)));
    }

    static void findAllDist(int planetCount) {
        for(int mid=0; mid<planetCount; mid++) {
            for(int start=0; start<planetCount; start++) {
                for(int end=0; end<planetCount; end++) {
                    planetDist[start][end] =
                            Math.min(planetDist[start][mid] + planetDist[mid][end], planetDist[start][end]);
                }
            }
        }
    }

    static int findMinDist(int cur, int visited) {
        if(visited == ALL) return 0;
        if(dp[cur][visited] != -1) return dp[cur][visited];
        int minCost = INF;
        for(int next=0; next<planetCount; next++) {
            if((visited & (1 << next)) != 0) continue;
            int cost =
                    planetDist[cur][next] +
                            findMinDist(next, visited | (1 << next));
            minCost = Math.min(minCost, cost);
        }

        return dp[cur][visited] = minCost;
    }
}

File: python/src/jan/week5/boj/우주탐사선_박재환.py
def solution():
    import sys
    input = sys.stdin.readline

    INF = 10 * 1000 + 5
    planet_count, launch_planet = map(int, input().split())
    planet_dist = [[INF] * planet_count for _ in range(planet_count)]
    for i in range(planet_count):
        arr = list(map(int, input().split()))
        for j in range(planet_count):
            planet_dist[i][j] = arr[j]

    def find_all_dist():
        for mid in range(planet_count):
            for start in range(planet_count):
                for end in range(planet_count):
                    planet_dist[start][end] = (
                        min(planet_dist[start][end],
                            planet_dist[start][mid] + planet_dist[mid][end])
                    )
    # 플로이드 워셜로 모든 경우의 수 구하기
    find_all_dist()
    ALL = (1 << planet_count) - 1
    dp_arr = [[-1] * (1 << planet_count) for _ in range(planet_count)]
    def find_min_cost(cur, visited):
        if visited == ALL:
            return 0
        if dp_arr[cur][visited] != -1:
            return dp_arr[cur][visited]
        min_cost = INF
        for next in range(planet_count):
            if (visited & (1 << next)) == 0:
                cost = (planet_dist[cur][next] +
                        find_min_cost(next, (visited | (1 << next))))

                min_cost = min(min_cost, cost)

        dp_arr[cur][visited] = min_cost
        return dp_arr[cur][visited]

    return find_min_cost(launch_planet, (1 << launch_planet))

if __name__ == '__main__':
    print(solution())